Lib/test/test_inspect/test_inspect.py

```diff
@@ -4623,6 +4623,96 @@ class D2(D1):
+    def test_from_frame(self):
+        ns = {}
+        def inner(a, /, b, *e, c: int = 3, d, **f) -> None:
+            ns['fr'] = inspect.currentframe()
+
+        inner(1, 2, d=4)
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '(a=1, /, b=2, *e, c=3, d=4, **f)')
+
+    def test_from_frame_with_pos_only_defaults(self):
+        ns = {}
+        def inner(a=1, /, b=2, *e, c: int = 3, d, **f) -> None:
+            ns['fr'] = inspect.currentframe()
+
+        inner(d=4)
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '(a=1, /, b=2, *e, c=3, d=4, **f)')
+
+    def test_from_frame_no_locals(self):
+        ns = {}
+        def inner():
+            ns['fr'] = inspect.currentframe()
+
+        inner()
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '()')
+
+    def test_from_frame_no_pos(self):
+        ns = {}
+        def inner(*, a, b=2, **c):
+            ns['fr'] = inspect.currentframe()
+
+        inner(a=1)
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '(*, a=1, b=2, **c)')
+
+    def test_from_frame_no_kw(self):
+        ns = {}
+        def inner(a, /, b, *c):
+            ns['fr'] = inspect.currentframe()
+
+        inner(1, 2)
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '(a=1, /, b=2, *c)')
+
+    def test_from_frame_with_nonlocal(self):
+        fr = None
+        def inner(a, /, b, *c):
+            nonlocal fr
+            fr = inspect.currentframe()
+
+        inner(1, 2)
+        self.assertEqual(str(inspect.Signature.from_frame(fr)),
+                         '(a=1, /, b=2, *c)')
+
+    def test_from_method_frame(self):
+        ns = {}
+        class _A:
+            def inner(self, a, *, b):
+                ns['fr'] = inspect.currentframe()
+            def __repr__(self):
+                return '_A'
+
+        _A().inner(1, b=2)
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '(self=_A, a=1, *, b=2)')
+
+    def test_from_frame_defaults_change(self):
+        ns = {}
+        def inner(a=1, /, c=5, *, b=2):
+            a = 3
+            ns['fr'] = inspect.currentframe()
+            b = 4
+
+        inner()
+        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
+                         '(a=3, /, c=5, *, b=4)')
+
+    def test_from_frame_mod(self):
+        self.assertEqual(str(inspect.Signature.from_frame(mod.fr)),
+                         '(x=11, y=14)')
+        self.assertEqual(str(inspect.Signature.from_frame(mod.fr.f_back)),
+                         '(a=7, /, b=8, c=9, d=3, e=4, f=5, *g, **h)')
+
+    def test_from_not_frame(self):
+        with self.assertRaisesRegex(TypeError, 'Frame object expected'):
+            inspect.Signature.from_frame(lambda: ...)
+
```

Comment:
Please add a test on a frame which has no locals:

```py
    def test_clear_frame(self):
        ns = {}
        def inner(a=1, /, c=5, *, b=2):
            ns['fr'] = inspect.currentframe()

        inner()
        ns['fr'].clear()
        self.assertEqual(str(inspect.Signature.from_frame(ns['fr'])),
                         '(a, /, c, *, b)')
```

We might raise an exception, but honestly, I prefer to have a signature without default values, rather than getting an exception.